Features assertion pass hook

Sup3rGeo opened the pull request:
Implements #3457.

Already has documentation and one simple test. Pending questions:

- Three pytest tests are not passing: 
`FAIL testing/acceptance_test.py::test_fixture_values_leak`
`FAIL testing/python/raises.py::TestRaises::()::test_raises_cyclic_reference[function]`
`FAIL testing/python/raises.py::TestRaises::()::test_raises_cyclic_reference[with]`
Need help to understand why they are failing.

- Always generating explanations (which could be expensive): As @nicoddemus mentionned this should be avoided BUT if assertion fails, we will generate them to report the failure. If the assert passes, users of the hook need it anyways, so it needs to be generated.
   - The only alternative I see is to not generate it if hook is not implemented by any plugins. Can we detect it? If possible, do you think it is worth it? 

- Is this test enough?

Thanks!

nicoddemus commented:
Quickly used [pytest-ast-back-to-python]() to see the rewritten code for this test:

```python
def test_hook():
    x = 1
    y = {2: 3}
    assert x == 1
```

**`Sup3rGeo:features-assertion-pass-hook`**

```
====================== Rewritten AST as Python ======================
import builtins as @py_builtins
import _pytest.assertion.rewrite as @pytest_ar
import sys
import pytest

def test_hook():
    x = 1
    y = {2: 3}
    @py_assert2 = 1
    @py_assert1 = x == @py_assert2
    @py_format4 = @pytest_ar._call_reprcompare(('==',), (@py_assert1,), ('%(py0)s == %(py3)s',), (x, @py_assert2)) % {'py0': @pytest_ar._saferepr(x) if 'x' in @py_builtins.locals() or @pytest_ar._should_repr_global_name(x) else 'x', 'py3': @pytest_ar._saferepr(@py_assert2)}
    @py_format6 = '%(py5)s' % {'py5': @py_format4}
    if not @py_assert1:
        raise AssertionError('' + ('assert ' + @pytest_ar._format_explanation(@py_format6)))
    else:
        @pytest_ar._call_assertion_pass(8, 'x == 1', @pytest_ar._format_explanation(@py_format6))
    @py_assert1 = @py_assert2 = None
    @py_format4 = @py_format6 = None
```

**`master`**

```
====================== Rewritten AST as Python ======================
import builtins as @py_builtins
import _pytest.assertion.rewrite as @pytest_ar
import sys
import pytest

def test_hook():
    x = 1
    y = {2: 3}
    @py_assert2 = 1
    @py_assert1 = x == @py_assert2
    if not @py_assert1:
        @py_format4 = @pytest_ar._call_reprcompare(('==',), (@py_assert1,), ('%(py0)s == %(py3)s',), (x, @py_assert2)) % {'py0': @pytest_ar._saferepr(x) if 'x' in @py_builtins.locals() or @pytest_ar._should_repr_global_name(x) else 'x', 'py3': @pytest_ar._saferepr(@py_assert2)}
        @py_format6 = ('' + 'assert %(py5)s') % {'py5': @py_format4}
        raise AssertionError(@pytest_ar._format_explanation(@py_format6))
    @py_assert1 = @py_assert2 = None
```

Sup3rGeo commented:
@nicoddemus no problem, I am glad you are here to help!

As you can see, the changes in the AST are relatively trivial. I guess the presence of the `pytest_ar` calls even when test continues normally could be the culprit?

> As an alternative, we should only activate this function if a certain config option is active, say assertion_pass=True

This might work as well, but there could be any confusion if a plugin implements the hook and the config option is not active? 
In other terms, do you see any drawbacks in making it automatically active if a plugin implements the hook? This would solve this possible confusion of manually setting the config.

nicoddemus commented:
@Sup3rGeo sorry for the delay on this.

Do you still have interest on this feature?

One of my concerns was the cost of always creating the explanation, but one way of solving that would be to pass a *function* to the hook that when called produces the explanation; it is weird in the sense that no hooks right now do that, but at least that would not incur a cost for normal usage.

Sup3rGeo commented:
Hi @nicoddemus, no problem at all

Yes I am still interested and I think your idea is good. Could also be an object that would hold a state in case we have many plugins implementing the hook, so it creates the explanation in the first call and stores it to be returned in any following call. And that does not format the explanation at all if no one uses it. I will try to come up with something when I have time.

However the thing I really cannot seem to do is to understand this leak and cyclic reference failures. Any idea how to approach this?

nicoddemus commented:
> However the thing I really cannot seem to do is to understand this leak and cyclic reference failures. Any idea how to approach this?

Let's see if using the lazy approach gets rid of the leaks, as I suspect they might.

Sup3rGeo commented:
One more interesting case: I changed `test_raises_cyclic_reference`, removing the `assert` keywords and using checking/raising `AssertionError` manually:

```
    def test_raises_cyclic_reference(self, method):
        """
        Ensure pytest.raises does not leave a reference cycle (#1965).
        """
        import gc

        class T(object):
            def __call__(self):
                raise ValueError

        t = T()
        if method == "function":
            pytest.raises(ValueError, t)
        else:
            with pytest.raises(ValueError):
                t()

        # ensure both forms of pytest.raises don't leave exceptions in sys.exc_info()
        if not sys.exc_info() == (None, None, None):
            raise AssertionError("NANA1")
        #locals()
        del t

        # ensure the t instance is not stuck in a cyclic reference
        for o in gc.get_objects():
            if type(o) is T:
                raise AssertionError("NANA2")
```

This test passes.

However, if I just uncomment the `locals()` statement (before `del t`), then it fails. So this confirms that having the `locals()` at the test body is by itself enough to fail this test.

RonnyPfannschmidt commented on src/_pytest/assertion/rewrite.py:
```diff
@@ -412,6 +412,10 @@ def _rewrite_test(config, fn):
     try:
+        import astunparse
+
+        with open(fn + "UNPARSED", "w", encoding="utf-8") as file:
```
is the unconditional unparse there for debugging?

Sup3rGeo replied:
Yes it was but already removed

Sup3rGeo commented:
One answer seems to confirm that `locals()` dictionary stays alive until the stack frame dies:
https://stackoverflow.com/questions/52337367/locals-and-object-reference-in-python

Sup3rGeo commented:
Current errors on CI are mainly:
- Error with xdist in the `fixture_values_leak_test.py`: That's because this depends on one test function to be executed after the other. Can we make xdist ignore or just run it on one thread?
- `astunparse` module uses the `async` name that became a keyword in python 3.7. I will see if I can rewrite this PR without using astunparse, as it could take a while until the next release and it seems inelegant to use the module just for this matter.

Guys, about avoiding unnecessary explanation calculation, I think that the best way is to check if any plugins implement the hookspec. If none does, then we only explain for failures (as before).
I think `pluggy` does not have this option but it should not be very hard to add to it, I would be willing to add this feature.
We can also add option switch to permanently turn it off, independently of any plugins/hooks. 

I think it is going to make everything simpler than passing a lazy object, which would also make the API more weird to use.

What do you think @nicoddemus and @RonnyPfannschmidt?

Sup3rGeo commented:
I added this to Pluggy (https://github.com/pytest-dev/pluggy/pull/177) and updated the code in this PR to use it. It seems to work just as expected.

However I was checking this saving the unparsed ast manually with/without hooks and comparing the modified source. I have no idea how to test it to make it part of the unit tests. Ideas?

nicoddemus commented:
> Oh well, now I just realized that having the AST code dependent also on the hook implementation by the plugins will create some problems for caching

Yeah... no ideas come to mind at the moment to handle that. 😕

nicoddemus commented:
Thanks for bringing this to attention again!

Most of the concerns here are regarding performance. How about if this rewriting is opt-in,  controlled by an `ini` option? This way most users will remain unaffected, and projects that need this can enable it. 

If plugins need this enabled, they can initially document so, but potentially in the future we can provide a way to enable this programmatically in a way that those plugins can do it at runtime.

Finally, we should mark the entire feature as experimental so we can revert it in case we discover some serious problem with it once it is in the wild.

Sup3rGeo commented:
What I am doing now is to check whether any plugins implement the `pytest_assertion_pass` hook (in the `_check_if_assertionpass_impl` function), and only format explanation and call the hook if so:

```
        if not @py_assert6:
            raise AssertionError(@pytest_ar._format_explanation('' + (
                'assert ' + @py_format11)))
        elif @pytest_ar._check_if_assertionpass_impl():
            @pytest_ar._call_assertion_pass(260,
                'len(spy_factory.instances) == 1', @pytest_ar.
                _format_explanation(@py_format11))
```

Would that be a reasonable solution?

nicoddemus commented:
> Would that be a reasonable solution?

Oh that is a good approach. I would still mark the hook as experimental thought.

@Sup3rGeo do you need this support in Python 2? I ask because we are about to release 4.6, which is supposed to be the last version supporting Python 2.

nicoddemus commented on setup.py:
```diff
@@ -17,6 +17,7 @@
     "pluggy>=0.12,<1.0",
     "importlib-metadata>=0.12",
     "wcwidth",
+    "astor",
```
Is this really needed, or can we work around it in someway? I would rather not having a new dependency unless strictly necessary.

Sup3rGeo replied:
Basically this is needed to provide the hook users with the python code that contains the assertion, and astor is used to go from the ast to python source code.

I would also prefer not to add dependencies, but I am not sure if there is anything already in place that does that.

Sup3rGeo replied:
@nicoddemus what about we incorporate the `codegen` module from `pytest-ast-back-to-python`? Or maybe the whole plugin?

nicoddemus replied:
Not sure, it has really limited scope (only for developers), but perhaps in the future? 🤔

Sup3rGeo replied:
Ok, so staying with `astor` as a dependency.

nicoddemus commented on src/_pytest/assertion/rewrite.py:
```diff
@@ -833,26 +849,47 @@ def visit_Assert(self, assert_):
         else:
             raise_ = ast.Raise(exc, None, None)
-        body.append(raise_)
+        # if self._assertion_pass:
```
Can we even skip meddling with the AST if there are no hookimpls implementing `pytest_assertion_pass`? I fear bugs on the assertion rewriting code will sprung once this is used in the wild and we might need to revert this feature in a rush like we did recently with the `all()` unrolling.

Sup3rGeo replied:
Rewriting conditional on hookimpls were my original idea but it might cause problem with caching, as discussed in the beginning: Users would need to clear the pyc cache everytime it adds a plugin that implements it or uses it himself in a `conftest.py`.

This might be a big problem or not, what do you think?

Sup3rGeo replied:
We could do a assertion rewriting switch like this but based on a command line option, which could be disabled by default while this hook is experimental (we can document that and even detect the hookimpls that this will only work if they enable via command line)

nicoddemus replied:
> Rewriting conditional on hookimpls were my original idea but it might cause problem with caching, as discussed in the beginning: Users would need to clear the pyc cache everytime it adds a plugin that implements it or uses it himself in a conftest.py.

I think this will have few uses initially, so actually I'm fine with this option; we can document it very prominently in the docs that this is necessary on activation.

To reiterate, I think this feature might spring some cool plugins, but I fear that this might affect users that don't use the hook at all, and assertion rewriting is pretty tricky as you know. 😁

Sup3rGeo commented:
> do you need this support in Python 2? I ask because we are about to release 4.6, which is supposed to be the last version supporting Python 2.

Probably not!

Sup3rGeo commented:
Is it normal that the module rewrites hypothesis from site-packages? Happens even on features:

```
Rewriting d:\work\pytest\src\_pytest\pytester.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\_settings.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\errors.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\compat.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\reflection.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\configuration.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\vendor\pretty.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\validation.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\coverage.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\utils\conventions.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\utils\dynamicvariables.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\control.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\reporting.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\core.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\strategies.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\_strategies.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\cache.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\cathetus.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\charmap.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\conjecture\utils.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\floats.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\entropy.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\searchstrategy\strategies.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\lazyformat.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\conjecture\data.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\conjecture\junkdrawer.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\escalation.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\searchstrategy\collections.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\searchstrategy\datetime.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\searchstrategy\deferred.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\searchstrategy\functions.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\searchstrategy\lazy.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\searchstrategy\misc.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\searchstrategy\numbers.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\conjecture\floats.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\searchstrategy\recursive.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\searchstrategy\shared.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\searchstrategy\strings.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\intervalsets.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\types.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\executors.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\conjecture\engine.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\conjecture\datatree.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\conjecture\shrinker.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\conjecture\shrinking\floats.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\conjecture\shrinking\common.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\conjecture\shrinking\integer.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\conjecture\shrinking\lexical.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\conjecture\shrinking\ordering.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\healthcheck.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\statistics.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\version.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\extra\pytestplugin.py
Rewriting D:\WORK\pytestenv36\lib\site-packages\hypothesis\internal\detection.py
Rewriting D:\WORK\pytest\testing\conftest.py
```

nicoddemus commented:
As per https://github.com/pytest-dev/pytest/pull/5373#issuecomment-504551231, please rebase this on features as we will be releasing `5.0` from there. 👍

nicoddemus commented:
Hi @Sup3rGeo, thanks for pursuing this! 😁 

It seems there's failing test and linting.

(You can `pip install pre-commit && pre-commit install` to have those checks done at each commit)

nicoddemus commented on src/_pytest/assertion/rewrite.py:
```diff
@@ -777,24 +793,46 @@ def visit_Assert(self, assert_):
+        else:
+            raise_ = ast.Raise(exc, None, None)
+        # Call to hook when passes
+        orig = astor.to_source(assert_.test).rstrip("\n").lstrip("(").rstrip(")")
```
I really would feel better if this assertion rewriting part could be made opt-in through an option.

Sup3rGeo replied:
You think the risk lies when calling the assertion pass hook or simply on the assertion rewriting changes?

In other words, should this switch just omit the `if` that calls the hook (but keeping the structural changes compared to before), or the switch should actually revert the entire rewriting to be exactly the same as before?

nicoddemus replied:
> You think the risk lies when calling the assertion pass hook or simply on the assertion rewriting changes?

The latter. 👍 

> In other words, should this switch just omit the if that calls the hook (but keeping the structural changes compared to before), or the switch should actually revert the entire rewriting to be exactly the same as before?

You got it, to revert the entire rewriting to be the same as before.

My fear is that, because there's a ton of code in the wild, things can go wrong in the assertion code itself, blowing up for users which don't even use the feature. This might be the kind of "major breakage" which I wouldn't like to see in the 5.0 release, hence I would prefer we skip the entire rewriting if the option is not set.

How hard you think it would be to include this switch? If it would complicate the code considerably we can skip it, but if it is simple to implement it, we should do it. 😁 

If you end up including the option, make sure to mention it in the docstring for the hook, so it is clear that users should set it for the hook to work.

As for the name of the option, I suggest `enable_pytest_assertion_pass_hook` defaulting to `False`, but feel to suggest something else.

Sup3rGeo replied:
Thanks! I added the option, although it was a little tricky to get it down to the assertion rewriter as the options are not yet parsed so we get it with `config.getvalue` it seems. You can take a look.

Also, one thing that might be problematic is that pytest rewrites and caches even site-packages libraries, so they should have the pyc files removed as well.

Sup3rGeo commented:
@nicoddemus concerning the following test:

https://github.com/pytest-dev/pytest/blob/a54e2e19f53eede956a0f8c8156bdb60ac40bce4/testing/acceptance_test.py#L1063-L1105

This one is failing because according to the changes I made, now the `pytest_assertrepr_compare` is always called even if the test passes - and it seems that having the objects passed for the `left` and `right` arguments is keeping a reference, so the test fails.

I thought about passing some sort of a weakref for the `left` and `right` arguments of the hook but you cannot really have a proxy for any kind of object.

I will keep digging but would you have any suggestions in the meantime?

nicoddemus commented:
I've added this to the 5.0 milestone, I'm confident it is complete enough to go into the 5.0 release. 🎉 

@Sup3rGeo let me know if you won't have time to finish this off in the next day or two, I will try to pick it up then. 👍

Sup3rGeo commented:
I guess by this point it should be mostly done.

nicoddemus commented:
Thanks @Sup3rGeo!

I noticed the default value for the option was `"False"` instead of `False`, which means the option was actually enabled by default as `"False"` is truthy, so I fixed that. That in turn uncovered a bug: the escape for `%` in `_format_assertmsg` was removed (probably by accident), so the original assertion rewrite code was left vulnerable to a bug. Also fixed. 👍 

I also took sometime to refine the docs and CHANGELOG a bit. 👍 

As soon as CI passes we can merge this I believe. 

Thanks a lot again, for your perseverance and patience under my barrage of requests! 😁

nicoddemus commented:
Pushed another small change: I moved the creation of the formatting variables when the assertion passes under the `if _check_if_assertion_pass_impl():` block:

```
import builtins as @py_builtins
import _pytest.assertion.rewrite as @pytest_ar

def test_foo():
    @py_assert0 = 1
    @py_assert3 = 2
    @py_assert2 = @py_assert0 == @py_assert3
    if @py_assert2 is None:
        from _pytest.warning_types import PytestAssertRewriteWarning
        from warnings import warn_explicit
        warn_explicit(PytestAssertRewriteWarning('asserting the value None, please use "assert is None"'), category=None, filename='c:\\pytest\\.tmp\\test-assert-foo.py', lineno=2)
    if not @py_assert2:
        @py_format5 = @pytest_ar._call_reprcompare(('==',), (@py_assert2,), ('%(py1)s == %(py4)s',), (@py_assert0, @py_assert3)) % {'py1': @pytest_ar._saferepr(@py_assert0), 'py4': @pytest_ar._saferepr(@py_assert3)}
        @py_format7 = '%(py6)s' % {'py6': @py_format5}
        raise AssertionError(@pytest_ar._format_explanation(@pytest_ar._format_assertmsg('To be escaped: %') + ('\n>assert ' + @py_format7)))
    elif @pytest_ar._check_if_assertion_pass_impl():
        @py_format5 = @pytest_ar._call_reprcompare(('==',), (@py_assert2,), ('%(py1)s == %(py4)s',), (@py_assert0, @py_assert3)) % {'py1': @pytest_ar._saferepr(@py_assert0), 'py4': @pytest_ar._saferepr(@py_assert3)}
        @py_format7 = '%(py6)s' % {'py6': @py_format5}
        @pytest_ar._call_assertion_pass(2, '1 == 2', @pytest_ar._format_explanation(@py_format7))
    @py_format5 = @py_format7 = None
    @py_assert0 = @py_assert2 = @py_assert3 = None
```

This is a small optimization which is important if/when we eventually drop the requirement of using the `enable_assertion_pass_hook` ini option.

nicoddemus commented:
🎉

Sup3rGeo commented:
Thanks @nicoddemus, I see there were a lot of tweaks!
Glad to have this merged after this lingering for more than a year :)
